core/storage/classroom/gae_models.py

```diff
@@ -32,6 +32,15 @@
 (base_models,) = models.Registry.import_models([models.Names.BASE_MODEL])
 
 
+class ImageDict(TypedDict, total=False):
```

Comment:
Thanks, naming of thumbnail_data and banner_data looks fine.

I also read your conversation with @nikitaevg about the image dict thing, and I want to point out a possible issue with storing these as a Json object in the storage layer -- it is actually going to make it harder to do migrations in the future. E.g. let's say we want to store the height of the image, and we add that as a new field to the json object. Then some datastore models have the old dict format and some have the new dict format, and it's hard to tell which is which. That is why, in our other models, we always have a schema_version property to go along with any JSONProperty dicts so that we know what sort of dict it is.

One thing you could therefore consider is separating the fields in the storage layer and only combining them into an object when the domain object is constructed. (That way, you can have a separate adapter function that takes the data and constructs the relevant ImageData object.)

I just wanted to bring this up before you make this decision final since once it is baked into the datastore it becomes hard to change. Any thoughts? /cc @nikitaevg too.

Reply:
Yes make sense. and I think image related data will not change very often so it makes sense to just have separate fields in storage layer than maintaining versions.

Reply:
🤔 

> Then some datastore models have the old dict format and some have the new dict format, and it's hard to tell which is which.

It's the same for the regular top-level field, isn't it? Like, I don't see the conceptual difference between adding a json field and adding a top-level field, you'll end up with "new" and "old" entries anyway and you'll have to migrate them.

I don't want to block Afzal on this, I don't have a strong opinion here as well, so yeah, let's proceed with a flat structure. I would appreciate it, @seanlip, if you explained this bit which I didn't understand.

Reply:
Sure thing -- basically, there are standard procedures built-in to App Engine datastore for adding fields. To add a field, you just add it, give it a default value, and if necessary run a job to populate it once the code is deployed.

With a top-level field, it's obvious which model instances have it populated or not, and this can be easily introspected and queried for in the datastore viewer. With a JSON field, you can't really see or query this internal structure easily and would need to manage all the details/logic of "does this new field exist or not for this model" in our code.